fdbserver/storageserver.actor.cpp

```diff
@@ -8746,6 +8951,622 @@ void AddingShard::addMutation(Version version,
+	if (g_network->isSimulated()) {
+		Optional<Value> pm = wait(data->storage.readValue(persistMoveInShardKey(shard->id())));
+		if (!pm.present()) {
+			TraceEvent(SevError, "UpdatedMoveInShardMetaDataNotFound", data->thisServerID)
+			    .detail("Shard", shard->toString())
+			    .detail("ShardKey", persistMoveInShardKey(shard->id()))
+			    .detail("DurableVersion", data->durableVersion.get());
+		}
+	}
+
+	data->storage.writeKeyValue(KeyValueRef(persistMoveInShardKey(shard->id()), moveInShardValue(*shard->meta)));
+	wait(data->durableVersion.whenAtLeast(data->storageVersion() + 1));
+	TraceEvent(SevDebug, "UpdatedMoveInShardMetaData", data->thisServerID)
+	    .detail("Shard", shard->toString())
+	    .detail("ShardKey", persistMoveInShardKey(shard->id()))
+	    .detail("DurableVersion", data->durableVersion.get());
+
+	return Void();
+}
+
+void changeServerKeysWithPhysicalShards(StorageServer* data,
+                                        const KeyRangeRef& keys,
+                                        const UID& dataMoveId,
+                                        bool nowAssigned,
+                                        Version version,
+                                        ChangeServerKeysContext context,
+                                        EnablePhysicalShardMove enablePSM);
+
+ACTOR Future<Void> fallBackToAddingShard(StorageServer* data, MoveInShard* moveInShard) {
+	if (moveInShard->getPhase() != MoveInPhase::Fetching) {
+		TraceEvent(SevError, "FallBackToAddingShardError", data->thisServerID)
+		    .detail("MoveInShard", moveInShard->meta->toString());
+		throw internal_error();
+	}
+	auto& mLV = data->addVersionToMutationLog(data->data().getLatestVersion());
+	TraceEvent(SevInfo, "FallBackToAddingShardBegin", data->thisServerID)
+	    .detail("Version", mLV.version)
+	    .detail("MoveInShard", moveInShard->meta->toString());
+	moveInShard->cancel();
+	for (const auto& range : moveInShard->meta->ranges) {
+		const Reference<ShardInfo>& currentShard = data->shards[range.begin];
+		if (currentShard->moveInShard && currentShard->moveInShard->id() == moveInShard->id()) {
+			ASSERT(range == currentShard->keys);
+			changeServerKeysWithPhysicalShards(data,
+			                                   range,
+			                                   moveInShard->dataMoveId(),
+			                                   true,
+			                                   mLV.version - 1,
+			                                   CSK_FALL_BACK,
+			                                   EnablePhysicalShardMove::False);
+		} else {
+			TraceEvent(SevWarn, "ShardAlreadyChanged", data->thisServerID)
+			    .detail("ShardRange", currentShard->keys)
+			    .detail("ShardState", currentShard->debugDescribeState());
+		}
+	}
+
+	wait(data->durableVersion.whenAtLeast(mLV.version + 1));
+
+	return Void();
+}
+
+ACTOR Future<Void> fetchShardCheckpoint(StorageServer* data, MoveInShard* moveInShard, std::string dir) {
```

Comment:
Folks will know how to use these methods? That they have to check for moveInShard->failed() on return?

Reply:
It is a bit unfortuante that we need to check `moveInShard->failed()` after each `wait` in all the `fetchShard*()` ACTORs, since the `ACTOR` cancellation is asynchronous for the child ACTORs created within `fetchShard()`.